File: tests/golden/java-spring-clean-multimodule/core/src/main/java/io/github/jtsato/walletservice/core/domains/wallet/usecase/create/CreateWalletUseCase.java
package io.github.jtsato.walletservice.core.domains.wallet.usecase.create;

import io.github.jtsato.walletservice.core.domains.wallet.model.Wallet;
import io.github.jtsato.walletservice.core.domains.wallet.usecase.create.CreateWalletCommand;

public interface CreateWalletUseCase {
    Wallet execute(CreateWalletCommand command);
}

File: tests/golden/nestjs-clean-architecture/core/src/core/usecases/create-wallet/create-wallet.usecase.ts
import { Wallet } from '../../models/wallet.model';
import { CreateWalletCommand } from './create-wallet.command';
import { CreateWalletCommandValidator } from './create-wallet-command.validator';
import { ICreateWalletGateway } from './create-wallet.gateway';
import { ICreateWalletUseCase } from './create-wallet-usecase.interface';

export class CreateWalletUseCase implements ICreateWalletUseCase {
  public constructor(
    private readonly gateway: ICreateWalletGateway,
    private readonly validator: CreateWalletCommandValidator = new CreateWalletCommandValidator(),
  ) {}

  public async execute(command: CreateWalletCommand): Promise<Wallet> {
    this.validator.validate(command);

    const wallet = new Wallet(
      command.id,
      command.balance,
    );

    return this.gateway.execute(wallet);
  }
}
